libs/@local/graph/migrations-macros/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+        ))
+    }
+}
+
+fn find_migration_files(
+    folder: impl AsRef<Path>,
+) -> impl Iterator<Item = Result<MigrationFile, ReadMigrationFileError>> {
+    WalkDir::new(folder)
+        .min_depth(1)
+        .max_depth(1)
+        .sort_by_file_name()
+        .into_iter()
+        .filter_map(Result::ok)
+        .filter(is_module_path)
+        .map(DirEntry::into_path)
+        .map(read_migration_file)
+}
+
+struct EmbedMigrationsInput {
+    location: PathBuf,
+}
+
+fn migration_list() -> TokenStream {
+    quote! {
+        struct Cons<H, T>(H, T);
+        struct Nil;
+
+        macro_rules! tuples {
+            () => { Nil };
+            ($head:expr) => { Cons($head, Nil) };
+            ($head:expr, $($tail:tt)*) => {
+                Cons($head, tuples![$($tail)*])
+            };
+        }
+
+        #[derive(Debug)]
+        pub struct MigrationDefinition<M> {
+            pub migration: M,
+            pub info: MigrationInfo,
+        }
+
+        impl<Head, Tail, C> MigrationList<C> for Cons<MigrationDefinition<Head>, Tail>
+        where
+            Head: Migration,
+            Tail: MigrationList<C>,
+            C: ContextProvider<Head::Context>,
+        {
+            async fn traverse(
+                self,
+                runner: &impl MigrationRunner,
+                context: &mut C,
+                direction: MigrationDirection,
+            ) -> Result<(), Report<MigrationError>> {
+                if direction == MigrationDirection::Down {
+                    self.1.traverse(runner, context, direction).await?;
+                    runner
+                        .run_migration(self.0.migration, &self.0.info, context.provide())
+                        .await
+                } else {
+                    runner
+                        .run_migration(self.0.migration, &self.0.info, context.provide())
+                        .await?;
+                    self.1.traverse(runner, context, direction).await
+                }
+            }
+        }
+
+        // Empty-list implementation.
+        impl<C> MigrationList<C> for Nil {
+            async fn traverse(
+                self,
+                _: &impl MigrationRunner,
+                _: &mut C,
+                _: MigrationDirection,
+            ) -> Result<(), Report<MigrationError>> {
+                Ok(())
+            }
+        }
+    }
+}
+
+fn embed_migrations_impl(input: &EmbedMigrationsInput) -> Result<TokenStream, Box<dyn Error>> {
+    let included_dir = input.location.to_string_lossy();
+    let mut migration_files = BTreeMap::new();
+    for migration_file in find_migration_files(&input.location) {
+        let migration_file = migration_file?;
+        match migration_files.entry(migration_file.number) {
+            Entry::Vacant(entry) => {
+                entry.insert(migration_file);
+            }
+            Entry::Occupied(entry) => {
+                return Err(ReadMigrationFileError::DuplicateMigration(
```

Comment:
have you thought about maybe using something like https://lib.rs/crates/proc-macro-error2 for better error locations?

Reply:
I didn't know about that crate. But changing the errors now is a bigger piece of work. I'm going to use it when I add more features such as storing a list of all migration infos which is required to detect missing migrations.